### ISSUE TITLE
Use Apple extensions to accelerate PBO data uploads

### ISSUE THREAD

**Issue by kvark** [servo/webrender#1874]
According to @jrmuizel , we spend 25% of CPU time in the `render` thread to move data over in an artificial test case. This is roughly split 3:1 between the PBO filling cost and PBO->GPU copy cost (which apparently involves `memmove` on Apple's GL...).

Using Apple's [recommended practices](https://developer.apple.com/library/content/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_texturedata/opengl_texturedata.html) would allow us to avoid some of the copies done by the driver, e.g.:
> GL_APPLE_client_storage allows you to prevent OpenGL from copying your texture data into the client. Instead, OpenGL keeps the memory pointer you provided when creating the texture. Your application must keep the texture data at that location until the referencing OpenGL texture is deleted.

**Comment by pcwalton** [comment]
Another possibility is to use IOSurface, which has a nicer API than that old extension. I suspect it's better maintained since it's used by Metal apps as well.

**Comment by jrmuizel** [comment]
Don't you need to Lock()/Unlock() with IOSurface?

Client storage allows DMAing directly out of client memory. The only really fast texture upload code that I've seen uses client storage. (i.e. that's what pre-metal CoreAnimation and the WindowServer use). Further, Metal seems to have persistently mapped buffers which avoid the Lock()/Unlock() of IOSurface.

**Comment by jrmuizel** [comment]
One issue that we may have with client storage though is texture formats. It's notoriously picky about things and easy to fall off the fast path.

**Comment by pcwalton** [comment]
Yeah, I remember using it in old Servo and hitting issues whereby we'd fall off the fast path with not only texture format but alignment and size of the buffer (which is why I switched to IOSurface). But that's definitely a good point; locking is not necessary with the client storage extension, so if we think we can reliably hit the fast paths it seems like the optimal way forward.

**Comment by jrmuizel** [comment]
See also https://github.com/servo/gleam/pull/126
